File: TreeCousins.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


#Time = O(n)
#Space = O(h)

class Solution:
    
    def dfs(self, root, depth, parent, x, y):
        
        if not root:
            return
        
        if root.val == x:
            self.parent_x = parent
            self.height_x = depth
            
            
        if root.val == y:
            self.parent_y = parent
            self.height_y = depth
            
            
        self.dfs(root.left, depth+1, root, x, y)
        self.dfs(root.right, depth+1, root, x, y)
    
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        self.parent_x = None
        self.parent_y = None
        
        self.height_x = 0
        self.height_y = 0
        
        self.dfs(root, 0, None, x, y)
        
        return self.parent_x != self.parent_y and self.height_x == self.height_y
        

File: treeCousins.java
//Time Complexity: O(n)
//Space Complexity: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        if(root == null) {
            return false;
        }

        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        boolean found;

        while(!q.isEmpty()) {
            found = false;
            int size = q.size();
            for(int i = 0; i < size; i++) {
                TreeNode curr = q.poll();

                if (curr.val == x || curr.val == y) {
                    if(found == false) {
                        found = true;
                    }
                    else {
                        return true;
                    }
                }

                if(curr.left != null && curr.right != null) {
                    if((curr.left.val == x && curr.right.val == y) || (curr.left.val == y && curr.right.val == x)) {
                        return false;
                    }
                }

                if(curr.left != null) {
                    q.add(curr.left);
                }
                if(curr.right != null) {
                    q.add(curr.right);
                }
            }

            if(found == true) {
                return false;
            }
        }

        return false;
    }
}